flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/AbstractExecutionSlotAllocator.java

```diff
@@ -105,22 +104,11 @@ protected SlotExecutionVertexAssignment createAndRegisterSlotExecutionVertexAssi
 		return slotExecutionVertexAssignment;
 	}
 
-	/**
-	 * Calculates the preferred locations for an execution.
-	 * It will first try to use preferred locations based on state,
-	 * if null, will use the preferred locations based on inputs.
-	 */
-	protected static CompletableFuture<Collection<TaskManagerLocation>> calculatePreferredLocations(
+	protected CompletableFuture<Collection<TaskManagerLocation>> calculatePreferredLocations(
```

Comment:
do we need this method `calculatePreferredLocations`?

Reply:
I think it's good to have to allow subclasses to not be aware of the way to retrieve preferred locations.

Reply:
Alright, then I believe `preferredLocationsRetriever` and `pendingSlotAssignments` fields can be private. Also the whole `AbstractExecutionSlotAllocator` can be package private and its protected constructor/methods. Any reason to keep them not package private?

Reply:
Yeah we can make it package private.
It is my oversight to make it public.

Reply:
This comment would be addressed in #12256 since I will rebase #12256 on this PR.